pkg/scanners/clairify/clairify.go

```diff
@@ -411,12 +420,44 @@ func (c *clairify) GetVulnerabilities(image *storage.Image, components *clairGRP
 	return convertImageToImageScan(image.GetMetadata(), resp.GetImage()), nil
 }
 
+func retryOnGRPCErrors(ctx context.Context, name string, f func() error) error {
+	op := func() error {
+		err := f()
+		if err != nil {
+			e, _ := status.FromError(err)
+			switch e.Code() {
+			// Unavailable indicates the service is currently unavailable. The error code is
+			// set by the gRPC framework during failed connections, abrupt shutdown of a
+			// server process or network connection.
+			//
+			// Aborted is not be generated by the gRPC framework, but may be used by server
+			// to retry on certain conditions that are retriable, e.g concurrency issue,
+			// sequencer check failures, transaction aborts, etc.
+			case codes.Aborted, codes.Unavailable:
+				return err
+			default:
+				return backoff.Permanent(err)
+			}
+		}
+		return err
+	}
+	notify := func(err error, duration time.Duration) {
+		log.Warnf("calling %s() (retrying in %s): %v", name, duration, err)
+	}
+	eb := backoff.NewExponentialBackOff()
+	return backoff.RetryNotify(op, backoff.WithContext(eb, ctx), notify)
+}
+
 func (c *clairify) GetNodeInventoryScan(node *storage.Node, inv *storage.NodeInventory) (*storage.NodeScan, error) {
 	req := convertNodeToVulnRequest(node, inv)
-	ctx, cancel := context.WithTimeout(context.Background(), clientTimeout)
+	ctx, cancel := context.WithTimeout(context.Background(), nodeScanClientTimeout)
```

Comment:
The timeout counter starts ticking in this line, right? From here, we got only 10s to make the call and potentially retry. I an unfortunate situation, the first call can take 10s and there will be no retries. 

Wouldn't it be better, to:
- Initialize the ctx with timeout of 10s only for a single call `c.nodeScanServiceClient.GetNodeVulnerabilities(ctx, req)` and
- configure the retry mechanism manually to make, e.g., max 3 attempts (or leave defaults and set the absolute max to 30s)?

Reply:
@vikin91:

> The timeout counter starts ticking in this line, right?

Yes.

> From here, we got only 10s to make the call and potentiall retries. I an unfortunate situation, the first call can take 10s and there will be no retries.

That's correct.

> Wouldn't it be better, to:

Potentially. It depends on which direction you want to improve.

I will explain why I chose that. Typically, if Scanner is timing out, something in its way is taking too long and is overloaded. In that scenario retrying multiple times might help increase pressure in Scanner. I experimented with Openshift-4, and `GetNodeInventoryScan` typically takes < 4s. Of course, we need more data to adjust the correct timeout. But ~2x seems reasonable to acknowledge Scanner is too busy, and we will release resources and try another time (later, like in 4hs if we don't send a retry message from Central to Compliance).

Reply:
Okay, I see that your intention was to squeeze the original call and 1-3 retries within the 10s window. This feels appropriate for our purposes, but does not let the exponential backoff to demonstrate its powers, as there will be no exponential behavior demonstrated for 1-3. 
This is by no means a blocker, I wanted to understand the behavior and design decisions.